File: experiments_realworld/dp/problem01/implementations/suboptimal_recursive/solution.py
import sys
sys.setrecursionlimit(1100000)

MOD = 1000000007

# SUBOPTIMAL: naive top-down recursion WITHOUT memoization.
# Same recurrence as the optimal (ways(s) = sum over coins of ways(s-coin)),
# but recomputes overlapping subproblems -> exponential in the answer size,
# while the memoized/iterative optimal is O(x*n). Computes the same value.
def solve(remaining, coins):
    if remaining == 0:
        return 1
    if remaining < 0:
        return 0
    result = 0
    for coin in coins:
        if remaining >= coin:
            result = (result + solve(remaining - coin, coins)) % MOD
    return result

def main():
    data = sys.stdin.buffer.read().split()
    it = iter(data)

    n = int(next(it))
    x = int(next(it))

    coins = []
    for _ in range(n):
        coins.append(int(next(it)))

    print(solve(x, coins))

if __name__ == "__main__":
    main()

File: experiments_realworld/dp/problem01/implementations/suboptimal_recursive/solution.cpp
#include <iostream>
#include <vector>
using namespace std;

const int MOD = 1000000007;

int n, x;
vector<int> coins;

// SUBOPTIMAL: naive top-down recursion WITHOUT memoization.
// Same recurrence as the optimal (ways(s) = sum over coins of ways(s-coin)),
// but recomputes overlapping subproblems -> exponential in the answer size,
// while the memoized/iterative optimal is O(x*n). Computes the same value.
int solve(int remaining) {
    if (remaining == 0) return 1;
    if (remaining < 0) return 0;

    int result = 0;
    for (int i = 0; i < n; i++) {
        if (remaining >= coins[i]) {
            result = (result + solve(remaining - coins[i])) % MOD;
        }
    }
    return result;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> n >> x;
    coins.resize(n);

    for (int i = 0; i < n; i++) {
        cin >> coins[i];
    }

    cout << solve(x) << "\n";

    return 0;
}
